### ISSUE TITLE
For a source code base, Identify key functions and generate missing unit tests

### ISSUE THREAD

**Issue by krishnapaparaju** [openshiftio/openshift.io#1792]
- [ ] Identify frequently used functions at a source code base (eg: GitHub repository)

- [ ] Flag if any of these functions do not have corresponding unit tests

- [ ] Generate source code for unit tests if unit tests are missing for these important functions

**Comment by krishnapaparaju** [comment]
For generating missing unit tests, please check out these links if we can reuse / collaborate with team at Rice University.

https://www.youtube.com/watch?v=e4CMTk-3iNk

http://pliny.rice.edu/index.html

**Comment by krishnapaparaju** [comment]
As part of understanding codebase, I have done a small experiment which converts a Java application code base into AST with https://github.com/antlr/antlr4. 'Antlr' does support rich grammars for other well known languages. Please see the attached screen shot capturing AST for the Java application I have tried. Once ASTs are in place, next step can be to figure out 'important functions' at a code base before start planning around AI based test case generation.
![antlr4_parse_tree_for_a_java_application](https://user-images.githubusercontent.com/15967765/34916870-bdecce3a-f964-11e7-8091-eaec16397b67.png)

**Comment by maxandersen** [comment]
@krishnapaparaju how are these generated tests supposed to be better than no tests ?

**Comment by krishnapaparaju** [comment]
@maxandersen this is an experiment to alert if unit tests are missing + auto generate unit tests

**Comment by krishnapaparaju** [comment]
Implementations around DeepCoder paper (https://openreview.net/pdf?id=ByldLrqlx) can be found at https://github.com/vaporized/DeepCoder-tensorflow & https://github.com/HiroakiMikami/deep-coder.

**Comment by maxandersen** [comment]
> @maxandersen this is an experiment to alert if unit tests are missing + auto generate unit tests

yes, I get that - but what value does it give the users to be told unit tests are missing vs having coverage tests running that would reveal the same ? 

what kind of unit tests are expected to be generated ? looking at paper it  seem to just be academic boundary tests ?  (which can be fine but it would be a big stretch to call those sufficient as unit tests)

**Comment by pkajaba** [comment]
@krishnapaparaju have you tried researching for existing projects? I stumbled upon this after short search https://github.com/randoop/randoop

**Comment by sivaavkd** [comment]
Moving this to Backlog
